## Graph

Below is provided a graph which vertices are different cities from Egypt and the edges are the distances to get from one city to another. We represent this graph with a dictionary because it is convenient.

In [4]:
graph = {
    'Alexandria' : [['Matruh', 159], ['Cairo', 112], ['Nekhel', 245]],
    'Matruh' : [['Siwa', 181], ['Alexandria', 159]],
    'Nekhel' : [['Alexandria', 245], ['Suez', 72], ['Quseir', 265]],
    'Cairo' : [['Alexandria', 112], ['Bawiti', 186], ['Asyut', 198]],
    'Siwa' : [['Matruh', 181], ['Bawiti', 210]],
    'Suez' : [['Nekhel', 72]],
    'Quseir' : [['Nekhel', 265], ['Sohag', 163]],
    'Bawiti' : [['Siwa', 210], ['Qasr Farafra', 104], ['Cairo', 186]],
    'Asyut' : [['Cairo', 198]],
    'Sohag' : [['Mut', 184], ['Qena', 69], ['Quseir', 163]],
    'Qasr Farafra' : [['Bawiti', 104], ['Mut', 126]],
    'Qena' : [['Sohag', 69], ['Luxor', 33]],
    'Mut' : [['Qasr Farafra', 126], ['Kharga', 98], ['Sohag', 184]],
    'Luxor' : [['Qena', 33]],
    'Kharga' : [['Mut', 98]]
}

## Straight Line Distance (SLD)

Additionally, two admissible heuristics h1 and h2, with distances from each city to Luxor are provided. We will also present both heuristics in a dictionary. Key: City, Value: [h1, h2]

In [5]:
sld = {
    'Alexandria' : [152, 163], ## I provided it for consistency
    'Matruh' : [174, 189],
    'Cairo' : [126, 139],
    'Nekhel' : [133, 145],
    'Siwa' : [132, 148],
    'Bawiti' : [105, 118],
    'Asyut' : [52, 67],
    'Suez' : [121, 136],
    'Qasr Farafra' : [68, 77],
    'Quseir' : [55, 59],
    'Mut' : [51, 65],
    'Kharga' : [24, 38],
    'Sohag' : [27, 36],
    'Qena' : [10, 19],
    'Luxor' : [0, 0]
}

## Queueing functions

The graph above will be traversed using two different algorithms:
* Greedy Search
* A* Search

The implementation of the algorithms only varies in the queueing function, so we designed an efficient way to implement the different traversings by a search function that calls a queueing function. 

The data structure used to do the queueing in each case is a dictionary. We will use two dictionaries: the first one represents the queue (priority queue) and the second one represents the nodes already visited. The key-value pair is `key: node`, `value: [path_cost, cost_to_goal, total_cost]`

Queueing function:

In [12]:
# Enqueues the children of a given start_node from a graph. It saves them in a priority queue 
# with their respective costs. Order_by decides which sorting to do for the queue, it could be 
# either 1: by cost_to_goal (greedy), or 2: by total_cost (A*).
def queueing(graph, start_node, start_node_costs, queue, visited, heuristic, order_by):
    for node, node_cost in graph[start_node]:
        path_cost = start_node_costs[0] + node_cost
        cost_to_goal = sld[node][heuristic-1]
        total_cost = path_cost + cost_to_goal
        if (node not in visited) and (node not in queue) and (node != None):
            queue[node] = [path_cost, cost_to_goal, total_cost]            
    visited[start_node] = start_node_costs
    queue = dict(sorted(queue.items(), key=lambda x: x[1][order_by]))
    return queue, visited

## Search Algorithm

Now we define our search function:

In [14]:
# Search function, traverses a graph until reaching the goal using a provided option of 
# Greedy search or A* search (Greedy, or A*). Start and goal of the graph are needed. Also,
# the heuristic to be used is needed (either 1 or 2). Queue sorting changes depending on 
# the option.
def search(graph, option, start_node, goal, heuristic):
    visited = {}
    queue = {}
    queue[start_node] = [0, sld[start_node][heuristic-1], 0 + sld[start_node][heuristic-1]]
    while queue:
        start_node = list(queue.keys())[0] # we will pop first in
        start_node_costs = queue.pop(start_node) # pop and save costs
        if start_node == goal:
            print(start_node, end = ".")
            print("\nDistance:", start_node_costs[0], "miles",
                  "\nNumber of Nodes Expanded", len(visited), "nodes")
            return
        print(start_node, end = ", ") # print visited node
        if option == "Greedy":
            queue, visited = queueing(graph, start_node, start_node_costs, queue, visited, heuristic, order_by = 1)
        elif option == "A*":
            queue, visited = queueing(graph, start_node, start_node_costs, queue, visited, heuristic, order_by = 2)


## Traversing Graph

Testing our search function on Egipt's map with the two different options:
* `'Greedy'`: Greedy Search
* `'A*'`: A* Search

In [16]:
options = ['Greedy', 'A*']
heuristics = [1, 2]
for heuristic in heuristics:
    print("\n---Heuristic", heuristic, "---")
    for option in options:
        print("\nResult using", option)
        print("Output: ", end = '')
        search(graph, option, 'Alexandria', 'Luxor', heuristic)


---Heuristic 1 ---

Result using Greedy
Output: Alexandria, Cairo, Asyut, Bawiti, Qasr Farafra, Mut, Kharga, Sohag, Qena, Luxor.
Distance: 814 miles 
Number of Nodes Expanded 9 nodes

Result using A*
Output: Alexandria, Cairo, Matruh, Asyut, Nekhel, Bawiti, Suez, Qasr Farafra, Siwa, Quseir, Mut, Kharga, Sohag, Qena, Luxor.
Distance: 775 miles 
Number of Nodes Expanded 14 nodes

---Heuristic 2 ---

Result using Greedy
Output: Alexandria, Cairo, Asyut, Bawiti, Qasr Farafra, Mut, Sohag, Qena, Luxor.
Distance: 814 miles 
Number of Nodes Expanded 8 nodes

Result using A*
Output: Alexandria, Cairo, Matruh, Asyut, Nekhel, Bawiti, Suez, Qasr Farafra, Siwa, Quseir, Mut, Kharga, Sohag, Qena, Luxor.
Distance: 775 miles 
Number of Nodes Expanded 14 nodes


As expected, A* got to an optimal solution for both heuristics. A* search reached the goal in a distance of 775 miles expanding 14 nodes for both heuristics. 

In the other hand, Greedy search traverses less city nodes, 9 for the first heuristic and 8 for the second heuristic, but it did not reach to an optimal solution, the distance found was 814 miles.

## Summary
### Greedy
|        |  Greedy( h1 )   |   Greedy( h2 )   |
|--------|---------|---------|
|**Distance**|814 miles|814 miles|
|**Nodes Expanded**|9 nodes|8 nodes|

### A*
|        |  A*( h1 )     |    A*( h2 )   |
|--------|---------|---------|
|**Distance**|775 miles|775 miles|
|**Nodes Expanded**|14 nodes|14 nodes|

## Additional Notes

### Inspecting each queue

In [17]:
def search(graph, option, start_node, goal, heuristic):
    visited = {}
    queue = {}
    queue[start_node] = [0, sld[start_node][heuristic-1], 0 + sld[start_node][heuristic-1]]
    while queue:
        print("\nqueue: ", queue)
        print("visited: ", visited, "\n")
        start_node = list(queue.keys())[0] # we will pop first in
        start_node_costs = queue.pop(start_node) # pop and save costs
        if start_node == goal:
            print(start_node, end = ".")
            print("\nDistance:", start_node_costs[0], "miles",
                  "\nNumber of Nodes Expanded", len(visited), "nodes")
            return
        print(start_node, end = ", ") # print visited node
        if option == "Greedy":
            queue, visited = queueing(graph, start_node, start_node_costs, queue, visited, heuristic, order_by = 1)
        elif option == "A*":
            queue, visited = queueing(graph, start_node, start_node_costs, queue, visited, heuristic, order_by = 2)


In [22]:
options = ['Greedy', 'A*']
heuristics = [1, 2]
for heuristic in heuristics:
    print("\n---Heuristic", heuristic, "---")
    for option in options:
        print("\nResult using", option)
        print("Output: ", end = '')
        search(graph, option, 'Alexandria', 'Luxor', heuristic)
        print("==============================================")


---Heuristic 1 ---

Result using Greedy
Output: 
queue:  {'Alexandria': [0, 152, 152]}
visited:  {} 

Alexandria, 
queue:  {'Cairo': [112, 126, 238], 'Nekhel': [245, 133, 378], 'Matruh': [159, 174, 333]}
visited:  {'Alexandria': [0, 152, 152]} 

Cairo, 
queue:  {'Asyut': [310, 52, 362], 'Bawiti': [298, 105, 403], 'Nekhel': [245, 133, 378], 'Matruh': [159, 174, 333]}
visited:  {'Alexandria': [0, 152, 152], 'Cairo': [112, 126, 238]} 

Asyut, 
queue:  {'Bawiti': [298, 105, 403], 'Nekhel': [245, 133, 378], 'Matruh': [159, 174, 333]}
visited:  {'Alexandria': [0, 152, 152], 'Cairo': [112, 126, 238], 'Asyut': [310, 52, 362]} 

Bawiti, 
queue:  {'Qasr Farafra': [402, 68, 470], 'Siwa': [508, 132, 640], 'Nekhel': [245, 133, 378], 'Matruh': [159, 174, 333]}
visited:  {'Alexandria': [0, 152, 152], 'Cairo': [112, 126, 238], 'Asyut': [310, 52, 362], 'Bawiti': [298, 105, 403]} 

Qasr Farafra, 
queue:  {'Mut': [528, 51, 579], 'Siwa': [508, 132, 640], 'Nekhel': [245, 133, 378], 'Matruh': [159, 174, 33